File: old/src/main/java/gfg/ds/DLLNode.java
package gfg.ds;

/** Created by gakshintala on 4/22/16. */
public class DLLNode {
	public int val;
	public DLLNode next;
	public DLLNode prev;

	public DLLNode(int val) {
		this.val = val;
	}

	public DLLNode(int val, DLLNode next, DLLNode prev) {
		this.val = val;
		this.next = next;
		this.prev = prev;
	}

	@Override
	public String toString() {
		return String.valueOf(this.val);
	}
}

File: common/src/main/kotlin/ds/ll/DLLNode.kt
package ds.ll

data class DLLNode<T>(var value: T, var prev: DLLNode<T>? = null, var next: DLLNode<T>? = null) {

  fun toList(): List<T> = mutableListOf(value).also { it.addAll<T>(next?.toList() ?: emptyList()) }

  override fun toString(): String =
    when {
      next == null -> value.toString()
      else -> "$value <=> $next"
    }

  companion object {
    fun <T> of(values: List<T>): DLLNode<T>? =
      if (values.isEmpty()) null else ofNonEmptyList(values)

    private tailrec fun <T> ofNonEmptyList(
      values: List<T>,
      index: Int = values.lastIndex,
      next: DLLNode<T> = DLLNode(values[index], null, null)
    ): DLLNode<T> =
      when (index) {
        0 -> next
        else -> {
          val node = DLLNode(values[index - 1], null, next)
          next.prev = node
          ofNonEmptyList(values, index - 1, node)
        }
      }
  }
}
